numba/targets/gdb_hook.py

```diff
@@ -150,27 +169,63 @@ def init_gdb_codegen(cgctx, builder, signature, args,
+    stat = builder.call(pipe, (pipefds_ptr,))
+    pipe_failed = builder.icmp_signed('==', stat, int32_t(-1))
+    with builder.if_then(pipe_failed, likely=False):
+        msg = "Internal error: `pipe` failed."
+        cgctx.call_conv.return_user_exc(builder, RuntimeError, (msg,))
+
+    # pointers to the read and write end FDs
+    read_end = builder.gep(pipefds, [zero_i32t], inbounds=True)
+    write_end = builder.gep(pipefds, [one_i32t], inbounds=True)
+
+    # this writes a command like "close(write_end_fd)" into the text buffer that
+    # will be one of the const char * that goes into the execl call made to
+    # gdb with view of a command like `-ex "close(write_end_fd)"` being executed
+    stat = builder.call(snprintf, (callclose_str, intp_t(20), callclosefmt,
+                                   builder.load(write_end)))
+    invalid_write = builder.icmp_signed('>', stat, int32_t(20))
+    with builder.if_then(invalid_write, likely=False):
+        msg = "Internal error: `snprintf` buffer would have overflowed."
+        cgctx.call_conv.return_user_exc(builder, RuntimeError, (msg,))
+
     is_child = builder.icmp_signed('==', child_pid, zero_i32t)
     with builder.if_else(is_child) as (then, orelse):
         with then:
             # is child
+
+            # first close the fork-inherited pipe ends, doesn't matter if they
+            # return something
+            builder.call(close, (builder.load(write_end),))
+            builder.call(close, (builder.load(read_end),))
+
+            # stage the gdb call
             nullptr = ir.Constant(char_ptr, None)
             gdb_str_ptr = builder.gep(
                 gdb_str, [zero_i32t], inbounds=True)
             attach_str_ptr = builder.gep(
                 attach_str, [zero_i32t], inbounds=True)
+            callclose_str_ptr = builder.gep(
+                callclose_str, [zero_i32t], inbounds=True)
             cgutils.printf(
                 builder, "Attaching to PID: %s\n", pidstr)
-            buf = (
-                gdb_str_ptr,
-                gdb_str_ptr,
-                attach_str_ptr,
-                pidstr_ptr)
+            buf = (gdb_str_ptr,    # gdb binary w path
+                   gdb_str_ptr,    # gdb binary w path
+                   attach_str_ptr, # the "attach" command
+                   pidstr_ptr,     # PID to attach to (parent, from above)
+                   cgctx.insert_const_string(mod, '-ex'), # str "-ex"
+                   callclose_str_ptr, # the command "close(write_end_fd)"
+                   )
+            cgutils.printf(builder,"%s\n", callclose_str)
             buf = buf + tuple(cmdlang) + (nullptr,)
             builder.call(execl, buf)
         with orelse:
             # is parent
-            builder.call(sleep, (int32_t(10),))
+            # issue a call to read() from the read-end of the pipe, this will
+            # block
+            builder.call(read, (builder.load(read_end), read_buffer, intp_t(1)))
```

Comment:
I think the parent that was blocked on the read(2) then needs to close(2) the read FD of the pipe to avoid leaking a file descriptor. Doesn't matter much if it's just one, but I don't know if this dance with gdb can re-occur. Also, the write FD of the pipe isn't needed in the parent and can be closed after the fork(2).